AppCenterIdentity/AppCenterIdentity/Internals/Storage/MSKeychainAuthTokenStorage.m

```diff
@@ -0,0 +1,128 @@
+  if (!tokenArray || tokenArray.count == 0) {
+    return nil;
+  }
+
+  MSAuthTokenInfo *authTokenInfo = tokenArray.firstObject;
+  NSDate *nextChangeTime = tokenArray.count > 1 ? tokenArray[1].startTime : nil;
+  if ([authTokenInfo.endTime laterDate:nextChangeTime]) {
+    return [[MSAuthTokenInfo alloc] initWithAuthToken:authTokenInfo.authToken
+                                         andStartTime:authTokenInfo.startTime
+                                           andEndTime:nextChangeTime];
+  }
+  return authTokenInfo;
+   */
+}
+
+- (void)saveAuthToken:(nullable NSString *)authToken withAccountId:(nullable NSString *)accountId expiresOn:(nullable NSDate *)expiresOn {
+  @synchronized(self) {
+
+    // Read token array from storage.
+    NSMutableArray<MSAuthTokenInfo *> *authTokensHistory = [self authTokensHistoryState];
+    NSString *latestAuthToken = [authTokensHistory lastObject].authToken;
+    if ((latestAuthToken != authToken) &&
+        !(latestAuthToken && authToken && [latestAuthToken isEqualToString:(NSString * _Nonnull) authToken])) {
+
+      // If new token differs from the last token of array - add it to array.
+      MSAuthTokenInfo *newAuthToken = [[MSAuthTokenInfo alloc] initWithAuthToken:authToken andStartTime:[NSDate date] andEndTime:expiresOn];
+      [authTokensHistory addObject:newAuthToken];
+    }
+
+    // Cap array size at max available size const (deleting from beginning).
+    if ([authTokensHistory count] > kMSIdentityMaxAuthTokenArraySize) {
+      [authTokensHistory removeObjectAtIndex:0];
+    }
+
+    // Save new array.
+    [self storeAuthTokensHistoryState:authTokensHistory];
+    if (authToken && accountId) {
+      [MS_USER_DEFAULTS setObject:(NSString *)accountId forKey:kMSIdentityMSALAccountHomeAccountKey];
+    } else {
+      [MS_USER_DEFAULTS removeObjectForKey:kMSIdentityMSALAccountHomeAccountKey];
+    }
+  }
+}
+
+// TODO: Finish the implementation of tokens removal as part of the separate PR.
+- (void)removeAuthToken:(nullable NSString *)__unused authToken {
+  /*
+  @synchronized(self) {
+
+    // Read token array from storage.
+    NSMutableArray<MSAuthTokenInfo *> *tokenArray = [self authTokensHistoryState];
+
+    // TODO: Allow only the oldest token to be removed.
+    // Do nothing if there's just one entry in the history or no history at all.
+    if (!tokenArray || tokenArray.count == 1) {
+      return;
+    }
+
+    // Find, delete the oldest entry. Do not delete the most recent entry.
+    for (NSUInteger i = 0; i < tokenArray.count - 1; i++) {
+      if ([tokenArray[i] authToken] == authToken) {
+        [tokenArray removeObjectAtIndex:i];
+        break;
+      }
+    }
+
+    // Save new array after changes.
+    [self storeAuthTokensHistoryState:tokenArray];
+  }
+   */
+}
+
+- (NSMutableArray<MSAuthTokenInfo *> *)authTokensHistoryState {
+  NSMutableArray<MSAuthTokenInfo *> *authTokensHistory = [MSKeychainUtil arrayForKey:kMSIdentityAuthTokenArrayKey];
```

Comment:
Cache it to avoid deserialization each time

Reply:
Sure. I've implemented it using property getter/setter. Please let me know if it looks good to you. Thanks.